### PR TITLE
[Linux][GDBJIT] Fix crashes related with locals debug info generation 

### PR DESCRIPTION
**Pull request by lucenticus**
This pull request fix several issues related with debug info generation for locals:
- Fix crash on generics static
- Add support of pointer type
- Fix incorrect processing of classes when we have several locals with the same class type

### CODE REVIEW

**Comment by lucenticus** [comment]
CC: @Dmitri-Botcharnikov @chunseoklee @seanshpark

**Comment by lucenticus** [comment]
@dotnet-bot test this please

**Comment by lucenticus** [comment]
@mikem8361 @janvorli Please take a look

**Comment by mikem8361** [comment]
LGTM but I really don't know the gdbjit code.

File: src/vm/gdbjit.cpp [hunk 1]
```diff
@@ -106,13 +107,15 @@ GetTypeInfoFromTypeHandle(TypeHandle typeHandle, NotifyGdb::PTK_TypeInfoMap pTyp
                 else
                 {
                     PTR_BYTE base = 0;
-                    if (!pField->IsRVA())
+                    MethodTable* pMT = pField->GetEnclosingMethodTable();
+                    base = pField->GetBaseInDomainLocalModule(pMT->GetDomainLocalModule(NULL));
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: This looks identical to pField->GetBase(), can we use it here?

Comment (lucenticus) [inline src/vm/gdbjit.cpp]: I tried to use this, but `GetBase()` doesn't exist if `DACCESS_COMPILE` is defined

Comment (janvorli) [inline src/vm/gdbjit.cpp]: Ah, I see.
Hmm, it is strange that it works for DACCESS_COMPILE whern you pass in NULL AppDomain parameter though. Looking at the MethodTable::GetDomainLocalModule(AppDomain * pAppDomain), I can see it calls GetModuleForStatics()->GetDomainLocalModule(pAppDomain); and the Module::GetDomainLocalModule(AppDomain *pDomain) asserts when DACCESS_COMPILE is defined and the AppDomain pointer is NULL (unless the first "if" in the function causes it to return before doing anything with the AppDomain pointer.

Comment (lucenticus) [inline src/vm/gdbjit.cpp]: It is strange, but it works (at least with my test samples:))  Probably, you could suggest better way to get base?

Comment (janvorli) [inline src/vm/gdbjit.cpp]: @noahfalk, @mikem8361 can you guys give an advice here?

**Comment by mikem8361** [comment]
When DACCESS_COMPILE is defined we are building libraries to go into mscordac which is used by the debugger code to access the VM data structures/code out of the target process. None of the gdbjit code doesn’t need to be compiled at when DACCESS_COMPILE is defined.  Even though it isn’t hurting anything currently to have all the extra code compiled into the DAC, it does make it bigger.  If you don’t want to make all of the gdbjit code conditional (on DACCESS_COMPILE), just “ifndef DACCESS_COMPILE” the call to GetBase().

**Comment by lucenticus** [comment]
@mikem8361 Thanks for clarifying! I've updated the PR, please check whether I understood you correctly.

**Comment by lucenticus** [comment]
@dotnet-bot  test  CentOS7.1 x64 Debug Build and Test please

**Comment by janvorli** [comment]
@mmitche there is something strange going on with the CentOS build:
```Error restoring build tools. This is either caused by a download error, or there is a locked file on the unix filesystem causing the failure ERROR: Could not restore build tools correctly. See '/mnt/resource/j/workspace/dotnet_coreclr/master/debug_centos7.1_prtest/init-tools.log' for more details.1```
Are you aware of that?

**Comment by mmitche** [comment]
@janvorli Nope.
@dotnet-bot test CentOS7.1 x64 Debug Build and Test

**Comment by mikem8361** [comment]
This will work for now, but what I really meant is that if DACCESS_COMPILE is defined, then none of the GDBJIT code should be enabled. When DACCESS_COMPILE is for the debugger dlls (mscordac) that are used out of proc for managed debugging and never “seem” by lldb/gdb.

**Comment by lucenticus** [comment]
@mikem8361  Ah, I've got it. I moved gdbjit.cpp file from VM_SOURCES_DAC_AND_WKS_COMMON to VM_SOURCES_WKS in CMakeLists.txt . Please take a look to the updated commit.

**Comment by lucenticus** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build and Test please

**Comment by mikem8361** [comment]
LGTM

**Comment by lucenticus** [comment]
Thank you!

**Pull request merged**